### ISSUE TITLE
Performing mutation with special characters on Aurora MySQL with GraphQL API returns with no error and no row is created on DB. 

### ISSUE THREAD

**Issue by rolisanchez** [awslabs/aws-mobile-appsync-sdk-ios#360]
**Describe the bug**

When performing a mutation using AppSync.perform(mutation:), with one of the properties of the mutation being a String with special characters (Symbols, Chinese characters and Emojis), the operation returns with no error, but the row is not created on the Aurora MySQL DB connected to my project via a GraphQL API. 

The problem is not on the DB, since performing a SQL INSERT Statement inserts the row correctly with special characters.  

I also tried using AppSync console on the browser and the mutation returns successfully and the row on the Relational DB is created.

**To Reproduce**
1. I created an Aurora MySQL DB on RDS Console. On the console, in Parameters Group, I proceeded to create a new DB cluster parameter group, with family set as aurora5.6. I changed the default value of `character_set_server` to utf8mb4. On the newly created DB, I went to Modify Cluster, then selected the DB cluster parameter group I just created, so the DB will have utf8mb4 as default and support special characters. 

After that, I proceeded to create the DB and Tables using the console: 

```
CREATE TABLE `User` (
    `userId` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
    `name` varchar(45) CHARACTER SET utf8mb4 NOT NULL DEFAULT 'no name',
    `email` varchar(45) COLLATE utf8mb4_bin DEFAULT NULL,
    `picture` varchar(45) CHARACTER SET utf8mb4 DEFAULT 'no picture',
    `createdAt` timestamp(6) NULL DEFAULT CURRENT_TIMESTAMP(6),
    `lastLoginAt` timestamp(6) NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6),
    PRIMARY KEY (`userId`),
    UNIQUE KEY `id_UNIQUE` (`userId`),
    UNIQUE KEY `email_UNIQUE` (`email`)
);

CREATE TABLE `Blog` (
    `blogId` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
    `blogName` varchar(45) CHARACTER SET utf8mb4 NOT NULL DEFAULT 'no name',
    `blogAuthorId` bigint(20) unsigned NOT NULL,
    PRIMARY KEY (`blogId`),
    UNIQUE KEY `blogId_UNIQUE` (`blogId`),
    KEY `blogAuthorId` (`blogAuthorId`),
    CONSTRAINT `blogAuthorId` FOREIGN KEY (`blogAuthorId`) REFERENCES `User` (`userId`) ON DELETE NO ACTION ON UPDATE NO ACTION
);
```

After creating the DB, I successfully tested that a new User can be created correctly with special characters on the console: 

`INSERT INTO User (name, email, picture) VALUES ("Johny's 很😄", "jdoe@gmail.com", "picture.png");`

I also tested with success, that the Foreign Key relationship creating a Blog for the newly created User, also on the console:

`INSERT INTO Blog (blogName, blogAuthorId) VALUES ("John’s Blog", 1);
`

After guaranteeing that my DB works accordingly, I proceeded to add an API on my Amplify iOS project, following this tutorial:

https://aws.amazon.com/blogs/mobile/aws-amplify-announces-new-amazon-aurora-serverless-and-graphql-transform-features-for-building-aws-appsync-apis/ 

Note that the selected API is GraphQL.

```
$ amplify add api
$ amplify api add-graphql-datasource
```
The Schema on AppSync is generated automatically during these CLI operations, and looks as follows (summarized):

```
type User {
	userId: Int!
	name: String
	email: String
	picture: String
	createdAt: AWSTimestamp
	lastLoginAt: AWSTimestamp
}

type Blog {
	blogId: Int!
	blogName: String!
	blogAuthorId: Int!
}

type Mutation {
	deleteBlog(blogId: Int!): Blog
	createBlog(createBlogInput: CreateBlogInput!): Blog
	updateBlog(updateBlogInput: UpdateBlogInput!): Blog
	deleteUser(userId: Int!): User
	createUser(createUserInput: CreateUserInput!): User
	updateUser(updateUserInput: UpdateUserInput!): User
}

input CreateUserInput {
	userId: Int!
	name: String
	email: String
	picture: String
	createdAt: AWSTimestamp
	lastLoginAt: AWSTimestamp
}
```

Note that the resolvers are also generated automatically.

I perform the following mutation successfully to test if the GraphQL API is working:

```
mutation {
  createUser(createUserInput: {
    userId: 18
    name: "Mike’s  很😄"
    email:"mike@gmail.com"
    picture:"empty"
  }) {
    name
  }
}
```

See screenshot below for the result.

In this point I know my DB is working with special characters and that the automatically generated Schemas are also working, so I proceeded to test this on my iOS App. (Note that the appSyncClient is already initialized in AppDelegate)

```
let createUserInput = CreateUserInput(userId: 2, name: "Triss 中文' 😄s", email: "trismerigold@gmail.com", picture: "picture.png")
let createUserMutation = CreateUserMutation(createUserInput: createUserInput)

let appSyncClient = (UIApplication.shared.delegate as! AppDelegate).appSyncClient

appSyncClient?.perform(mutation: createUserMutation) { (result, error) in
    if let error = error as? AWSAppSyncClientError {
        print("Error occurred: \(error)")
        print("Error occurred (localized): \(error.localizedDescription )")
        return
     }
     if let resultError = result?.errors {
         print("Error saving user: \(resultError)")
         return
      }
      print("result \(result)")
      print("(result) type \(type(of: result))")
      // Should return user here according to documentation.
      if let resultData = result?.data, let createdUser = resultData.createUser {
          print("User created: \(String(describing: createdUser.userId))")
      }
}
```

The printed result I got is:

`result Optional(AWSAppSync.GraphQLResult<SwiftUIAuroraAmplify.CreateUserMutation.Data>(data: Optional(SwiftUIAuroraAmplify.CreateUserMutation.Data(snapshot: ["createUser": nil])), errors: nil, source: AWSAppSync.GraphQLResult<SwiftUIAuroraAmplify.CreateUserMutation.Data>.Source.server, dependentKeys: Optional(Set(["MUTATION_ROOT.createUser([createUserInput:email:trismerigold@gmail.com,name:Triss 中文\' 😄s,picture:picture.png,userId:2)"]))))`

When checking the DB, no User was inserted.

If I try to insert a user without special characters, the User is inserted correctly, albeit the timestamp cannot be serialized from the result of the User mutation:

```
Unable to serialize `2020-03-11 13:47:16.188486` as a valid timestamp
```

The timestamp is the least of the problems though, since if I perform a query to list all users, the timestamp is serialized correctly.

**Expected behavior**
I expect the User with special characters on its name to be created successfully, as I know the DB accepts the characters and I can perform the mutation from the AppSync console successfully. 

**Screenshots**

<img width="1268" alt="aurora_graphql_mutation" src="https://user-images.githubusercontent.com/7276421/76422722-a5943880-63e0-11ea-806a-911cf1c74b7c.png">

**Environment(please complete the following information):**
 - AppSync SDK Version: (See below)
 - Dependency Manager: Cocoapods
 - Swift Version : 5

```
PODS:
  - AWSAppSync (3.0.2):
    - AppSyncRealTimeClient (~> 1.0.0)
    - AWSCore (~> 2.12.0)
    - ReachabilitySwift (~> 5.0.0)
    - SQLite.swift (~> 0.12.2)
```

**Device Information (please complete the following information):**
 - Device: iPhone 11 Pro, iPhone 11 Pro Max Simulator
 - iOS Version: iOS 13.3.1

**Additional context**

Chinese characters are a strict requirement in our project, since we are based in Taiwan.

I performed similar steps using DynamoDB and got the iOS mutation working with special characters on the name.

**Comment by Ricardo1980** [comment]
I have something similar and for me it is working fine.

Basically, there are 2 steps:

Regarding the database, RDS, go to "parameters group" and setup properly "utf8mb4" (I don't remember right now the specific steps). Check: https://stackoverflow.com/questions/49962291/aws-aurora-rds-mysql-in-a-cluster-cannot-insert-emoji
(not only character_set_server)

Secondly, your SQL creating script should be something like:
`nickname` varchar(64) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL
(I see I have utf8mb4_unicode_ci but you have something different)

It is working perfectly for me. My users can use emojis from the iPhone app, and this info is sent to Aurora serverless through AppSync/GraphQL without problems. And, of course, other users can read this data properly.

Try that, perhaps it works.

**Comment by rolisanchez** [comment]
> I have something similar and for me it is working fine.
> 
> Basically, there are 2 steps:
> 
> Regarding the database, RDS, go to "parameters group" and setup properly "utf8mb4" (I don't remember right now the specific steps). Check: https://stackoverflow.com/questions/49962291/aws-aurora-rds-mysql-in-a-cluster-cannot-insert-emoji
> (not only character_set_server)
> 
> Secondly, your SQL creating script should be something like:
> `nickname` varchar(64) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL
> (I see I have utf8mb4_unicode_ci but you have something different)
> 
> It is working perfectly for me. My users can use emojis from the iPhone app, and this info is sent to Aurora serverless through AppSync/GraphQL without problems. And, of course, other users can read this data properly.
> 
> Try that, perhaps it works.

Hi @Ricardo1980 . As I mentioned above, I already modified the parameters group `character_set_server `. The RDS DB is accepting the special characters when I send them from the AppSync console and when I perform SQL Statements such as `INSERT INTO User (name, email, picture) VALUES ("John’s  很😄", "jdoe222@gmail.com", "picture.png");`. 

As for your collate recommendation, I've run the following command on my RDS console: ` ALTER TABLE User MODIFY name varchar(45) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
`. 

As you can see in the image below, it ran succesfully.

<img width="1336" alt="Alter_Table" src="https://user-images.githubusercontent.com/7276421/76825459-04303b00-6855-11ea-9bc6-72ddf3d4d26f.png">

As before, no problem when inserting using `INSERT` in the console.

After that, I proceeded to regenerate my API, since I made some changes in the DB and I wanted to be sure it was re-done. I removed the API from my Amplify Project, then added it again, prompting the generation. (Following this to add Aurora API https://aws.amazon.com/blogs/mobile/aws-amplify-announces-new-amazon-aurora-serverless-and-graphql-transform-features-for-building-aws-appsync-apis/)

When testing the mutations on the AppSync console, now actually it **stopped** working and I receive the following: `createUser: null` (See image below)


<img width="904" alt="aws_appsync" src="https://user-images.githubusercontent.com/7276421/76827082-f1b80080-6858-11ea-90da-bf0c5ae572e0.png">

While now I can't create Users from the AppSync console like before, at least it's a behavior similar to the iOS App, in which I also get a response `["createUser": nil]`

![ios](https://user-images.githubusercontent.com/7276421/76827206-49566c00-6859-11ea-9e8c-c1d4f422e93a.png)

**Comment by rolisanchez** [comment]
UPDATE: If anyone is facing this problem. The problem was the single quotes character  ' I was trying to use not being correctly escaped.  This led to a badly formatted SQL Statement like:

`INSERT INTO User (name, email, picture) VALUES ('John Doe's', 'jdoe@gmail.com', 'picture.png');`

Notice the non escaped single quote in the string "John Doe's". Before finding out about this, I didn't escape it in iOS, since I'm using double quotes for a String and single quotes inside it, so I thought didn't need to escape it.

This might be a problem on the generated resolver when adding Aurora, since it's generating a SQL Statement in a String (https://github.com/aws/aws-appsync-community/issues/44). I tried editing the resolver to replace a single quote to an escaped single quote, but my knowledge of VTL wasn't good enough ... *sigh* ...  Meanwhile, I am doing escaping single and double quotes  in iOS before performing the mutation: 

```
let name = "Mithras's 中文 😄s".replacingOccurrences(of: "'", with: "''").replacingOccurrences(of: "\"", with: "\\\"")
let newUser = thisUser(userId: 33, name: name, email: "mithras@gmail.com", picture: "picture.png") 
```

TL;DR: Follow @Ricardo1980 recommendation to add `CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci ` and remember to escape any single/double quotes!

**Comment by Ricardo1980** [comment]
I had a similar problem with new lines (\n) and I solved it using $util.escapeJavaScript in my resolvers.
